jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpCookieStore.java

```diff
@@ -0,0 +1,388 @@
+//
+// ========================================================================
+// Copyright (c) 1995-2022 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.http;
+
+import java.net.URI;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+
+import org.eclipse.jetty.util.NanoTime;
+import org.eclipse.jetty.util.thread.AutoLock;
+
+/**
+ * <p>A container for {@link HttpCookie}s.</p>
+ * <p>HTTP cookies are stored along with a {@link URI} via {@link #add(URI, HttpCookie)}
+ * and retrieved via {@link #match(URI)}, which implements the path matching algorithm
+ * defined by <a href="https://www.rfc-editor.org/rfc/rfc6265">RFC 6265</a>.</p>
+ */
+public interface HttpCookieStore
+{
+    /**
+     * <p>Adds a cookie to this store, if possible.</p>
+     * <p>The cookie may not be added for various reasons; for example,
+     * it may be already expired, or its domain attribute does not
+     * match that of the URI, etc.</p>
+     * <p>The cookie is associated with the given {@code URI}, so that
+     * a call to {@link #match(URI)} returns the cookie only if the
+     * URIs match.</p>
+     *
+     * @param uri the {@code URI} associated with the cookie
+     * @param cookie the cookie to add
+     * @return whether the cookie has been added
+     */
+    public boolean add(URI uri, HttpCookie cookie);
+
+    /**
+     * @return all the cookies
+     */
+    public List<HttpCookie> all();
+
+    /**
+     * <p>Returns the cookies that match the given {@code URI}.</p>
+     *
+     * @param uri the {@code URI} to match against
+     * @return a list of cookies that match the given {@code URI}
+     */
+    public List<HttpCookie> match(URI uri);
```

Comment:
Would the following API be more efficient:
```suggestion
    public Iterator<HttpCookie> matches(URI uri);
```

As it would not need to build a list, rather just return a wrapper around the standard iterator.

Reply:
I'm not sure I understand.

I prefer `match()` because it is a command (so not `matches()` that looks like a query that should return a boolean).

The algorithm has to do a bunch of passes through the internal data structure to collect the matching cookies, so it is natural to return a list rather than an iterator (because it is where cookies are collected during the various passes).

The only use of this case is for the client to send the cookies on requests, so the list will be fully iterated (so a lazy iterator that may not be fully consumed is not useful).

Reply:
If multiple passes are done, then a simple wrapper of the internal iterator will not work and a collection is still needed.  So the current signature is good.